### PR TITLE
Added terraform and ci guidelines

### PR DESCRIPTION
**Pull request by ltshb**

### CODE REVIEW

File: TERRAFORM.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+## Code Organization
+
+The high level code organization should look like this
+
+```text
+|- modules/                   # terraform reusable modules for the whole repository
+|- ORGANIZATIONS_OR_ACCOUNT   # e.g. AWS Account name
+|           |- modules        # terraform reusable modules for this organization/account
+|           |- shared         # shared resources
+|           |- STAGING        # resource pro staging
```
Comment (boecklic) [inline TERRAFORM.md:32]: maybe `ENVIRONMENT` (staging is sometimes used instead of integration as name for an environment)

File: TERRAFORM.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+OLD_KEY=<old-state-key>
+
+aws --profile AWS_ACCOUNT_PROFILE s3 rm s3://STATE_BUCKET/${OLD_KEY}
+
+aws --profile AWS_ACCOUNT_PROFILE dynamodb delete-item --table-name TABLE_NAME --key '{"LockID": {"S": "STATE_BUCKET/'${OLD_KEY}'-md5"}}'
+```
+
+### Module Guidelines
+
+- :warning: ALWAYS changes resources using terraform (no manual changes via UI, e.g. AWS web console) :warning:
```
Comment (boecklic) [inline TERRAFORM.md:82]: `change`

File: TERRAFORM.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+### State Locking
+
+Enable state locking to prevent concurrent operations from corrupting your state.
+
+- Use AWS DynamoDB to manage state locks.
+
+### State migration
+
+To migrate the state we use the `terraform init --migrate-state` command. After migrating the state we need to manually clean up the old state by removing the S3 file and deleting the dynamoDB entry
```
Comment (boecklic) [inline TERRAFORM.md:70]: Should we give some more information here on how to migrate states?

**Review by boecklic — APPROVED** [review]
:+1:

File: CI.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+## Guidelines
+
+- Each Codebuild project is managed via terraform in a dedicated repository
+- `buildspec.yml` file is located within the terraform module of the Codebuild project which is in a private repository. This ensure security where a non authorized person cannot see and edit the buildspec and trigger a build on a public repository.
+- Don't use secrets/password inside the buildspec but use AWS SSM [AWS System Manager (SSM)](https://docs.aws.amazon.com/systems-manager/index.html)
+- Use build badge in the software github repository see [Build Badge](#build-badge)
+
+## Build Badge
+
+Build badge should be added to the top of the README.md file of the github project using the following template:
```
Comment (adk-swisstopo) [inline CI.md:25]: I suggest adding a link to https://docs.aws.amazon.com/codebuild/latest/userguide/sample-build-badges.htmlI didn't know what a build badge was until I looked it up. In fact the last paragraph in this section can probably be replaced with a link to https://docs.aws.amazon.com/codebuild/latest/userguide/sample-build-badges.html#access-badges

Also, can we add a linting rule that will complain if there is no build badge link in the README of a project?

File: TERRAFORM.md (hunk 1 above)
Comment (adk-swisstopo) [reply TERRAFORM.md]: +1 for ENVIRONMENT
Maybe also spell out some examples in the comment.

File: TERRAFORM.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+```text
+|- modules/                   # terraform reusable modules for the whole repository
+|- ORGANIZATIONS_OR_ACCOUNT   # e.g. AWS Account name
+|           |- modules        # terraform reusable modules for this organization/account
+|           |- shared         # shared resources
+|           |- STAGING        # resource pro staging
+```
+
+Then it should mirror the hierarchical structure of the infrastructure. Note that each terraform module (project) should be kept small.
```
Comment (adk-swisstopo) [inline TERRAFORM.md:35]: I am unclear about what the "hierachical structure of the infrastructure" means here.
Also, in order to enforce the "keep it small" constraint, should we make the CI complain when a module/project exceeds N lines of code?

File: TERRAFORM.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+### Reusable Module Usage
+
+Use modules to encapsulate and reuse configurations. This promotes DRY (Don't Repeat Yourself) principles and makes your configurations more manageable.
```
Comment (adk-swisstopo) [inline TERRAFORM.md:39]: We probably want to point out the caveats of modules here. The main one I think is that when you change a module, we don't currently have a good way to check what changes this triggers across the infrastructure and what needs to be pushed/applied.

File: TERRAFORM.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  - [Reusable Module Usage](#reusable-module-usage)
+- [State Management](#state-management)
+  - [Remote State](#remote-state)
+    - [State key](#state-key)
+  - [State Locking](#state-locking)
+  - [State migration](#state-migration)
+  - [Module Guidelines](#module-guidelines)
+- [Code formatting](#code-formatting)
+
+## Introduction
```
Comment (adk-swisstopo) [inline TERRAFORM.md:19]: Maybe we want a nomenclature section or point to the upstream documentation about this. I know I am still sometimes confused about how we use the words "module", "project" and "workspace" which I think is a little different from the way the official Terraform documentation uses them.

File: TERRAFORM.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+- Use modules to share global variales between terraform modules
+
+## State Management
+
+### Remote State
+
+We use remote state management using AWS S3 backend
+
+- Encrypt the state file at rest (see [Terraform Documentation](https://developer.hashicorp.com/terraform/language/settings/backends/s3#encrypt)).
+- Put the state backend definition in `terraform.tf` file
```
Comment (adk-swisstopo) [inline TERRAFORM.md:52]: Can we enforce that with the CI?

File: TERRAFORM.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+#### State key
+
+We use the terraform module definition file path as state key in S3 prefixed with a version string `v1`; e.g. `v1/infra-terraform-bgdi/systems-prod/map.terraform.tf`
+
+The version prefix is to avoid any risk of key collision in the future in case of code structure re-organization.
```
Comment (adk-swisstopo) [inline TERRAFORM.md:58]: Can we enforce correct key usage with the CI?

File: TERRAFORM.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+*NOTE: We used to use an UUID as state keys, which has the advantage that we did not have to change the state when reorganizing resources. However it has the downside that UUID key were very hard to code review and error prone due to copy paste mistake not being easily identifiable.*
+
+### State Locking
+
+Enable state locking to prevent concurrent operations from corrupting your state.
+
+- Use AWS DynamoDB to manage state locks.
```
Comment (adk-swisstopo) [inline TERRAFORM.md:66]: That sounds like "you need to use a special option but we won't tell you the exact commandline here". I think what it means is that we have a "dynamodb_table" field in the terraform.backend.s3 block. Can we just enforce that in the CI then we can delete this section? Alternatively, explain how it's done or link to an example.

File: TERRAFORM.md (hunk 3 above)
Comment (adk-swisstopo) [reply TERRAFORM.md]: For the "how", if the terraform documentation is good enough we can just link to that. However I think it would be good to document why we would want to migrate the state.

Comment (ltshb) [reply TERRAFORM.md]: done

File: TERRAFORM.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+aws --profile AWS_ACCOUNT_PROFILE s3 rm s3://STATE_BUCKET/${OLD_KEY}
+
+aws --profile AWS_ACCOUNT_PROFILE dynamodb delete-item --table-name TABLE_NAME --key '{"LockID": {"S": "STATE_BUCKET/'${OLD_KEY}'-md5"}}'
+```
+
+### Module Guidelines
+
+- :warning: ALWAYS change resources using terraform (no manual changes via UI, e.g. AWS web console) :warning:
+- :warning: NEVER EVER apply changes before opening a Pull Request :warning:
```
Comment (adk-swisstopo) [inline TERRAFORM.md:90]: Depending how you read this, I don't think that's how we currently do it, neither do I think that's really tractable. At least in dev. The workflow I have been using is:
1. prepare a first draft of my change, try to check it does the right thing with "terraform plan"
2. open a PR
3. maybe get someone to look at the PR, maybe have it approved
4.  "terraform apply" the change
5. fix what doesn't work, go back to 4
6. once it does the right thing, push my chanages to the PR
7. get the PR approved
8. merge the PR
9. maybe verify "terraform apply" in master doesn't show any change

In the cycle between 4 and 5, the state of the PR and the state of what I "terraform apply" with be inconsistent.

I think the important part is to make sure that anybody who runs into an inconsistent state needs to be able to easily track down what's going on. Hence, it is important to have a PR open that people can find even if its state is not 100% consistent all the time. And to not leave changes that have been terraform-applied but not github-merged hanging for too long. I think this change explains the last part but maybe it can do a better job at explaining the first part (the "why").

Also, maybe if we can have a bot that announce Terraform PRs on Slack that would be handy?

Also also, this whole section sounds more like "development guidelines" than "module guidelines".

File: TERRAFORM.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+- `master` branch SHOULD reflect the current infrastructure state, which means PR that are applied should be merged as soon as they have been reviewed. DON'T keep applied PR open for more that a few days.
+- Keep track of applied changes in the PR using the checkbox
+  - [ ] changes applied
```
Comment (adk-swisstopo) [inline TERRAFORM.md:93]: Initially I found little use for that checkbox and it's very easy to forget to tick/untick it consistently then it's just misleading. I think it becomes more useful when there are multiple "terraform apply" to perform for a given PR or other things to check/change along with it. I like how you suggest modifying/adding checkboxes as needed. It's also OK to remove some :)

File: TERRAFORM.md [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  
+  Modify/add checkbox if needed (e.g. one checkbox per module/staging, etc)
+- For AWS ressources we use one state bucket per AWS Account, we don't allow cross account state bucket access. This allow a simpler terraform state bucket access management and also improve security.
```
Comment (adk-swisstopo) [inline TERRAFORM.md:96]: Can we enforce the no-cross-account state sharing with the CI?

File: TERRAFORM.md [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+- Inside the same AWS account we can use remote state to access other terraform module data to avoid hardcoding
+- Avoid using magic number like AWS account id in your code but use variables, locals or global variables defined in reusable module instead
+- Avoid hard dependencies between modules whenever possible. By hard dependencies, I mean using remote state for a value when the use of the value in the resource doesn't need to exists, for example for AWS policies and role you might be tempted to get the role ARN or resource ARN from a remote state to create a new role or policy, but this would add a hard dependency on the terraform module where one module needs to be applied before the other, while on the AWS resource you don't have any hard dependencies, the ARN resource in a policy doesn't need to exists to create the policy.
```
Comment (adk-swisstopo) [inline TERRAFORM.md:99]: That sounds more like "unnecessary" than "hard" dependencies? Is there a way to avoid those without copy/pasting ARNs around? I am also unclear on how to determine whether a dependency is unnecessary/hard in that way without knowing much more than is reasonable about every AWS object type.

File: TERRAFORM.md [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+## Code formatting
+
+All terraform code should have the same formatting, for this we use the standard `terraform fmt` tool to format the code.
```
Comment (adk-swisstopo) [inline TERRAFORM.md:103]: Can we make sure the CI complains when the formatting is wrong? If it already does, we can mention it here.

**Review by adk-swisstopo — COMMENTED** [review]
Thank you for this.

I added a bunch of minor comments, some requests for clarifications and many suggestion to enforce things with the CI instead of documentation. For that last point, if it's not trivial I guess we could open a Jira task and point to that task (e.g. "remove/simplify this section once it's enforced automatically as per PB-...").

File: TERRAFORM.md (hunk 7 above)
Comment (ltshb) [reply TERRAFORM.md:19]: As I haven't read the terraform docu for a long while maybe could you summarize how you understand it as you have recently read it :smile:
We currently don't use terraform workspace functionality.

File: TERRAFORM.md (hunk 5 above)
Comment (ltshb) [reply TERRAFORM.md:35]: Here I wanted to be quite generic and don't provide any informations on our stack infrastructure because the guideline repository is public. You will find our stack hierarchy in our confluence pat.

About keeping it small you cannot really define a number of code line, you might have some "small" terraform module with a lots of line if you need to have configuration file, or if you require to have complex policies. So enforcing this by a CI would be quite difficult, I think the enforcement should be more in the review process, where the developer and the review should have some common sense to define when to split the code into several module.
For example on our stack we have a module that define the "system" aka Cloudfront with DNS and WAF, and have another terraform module that define the service itself (k8s namespace, bucket for service, etc).

File: TERRAFORM.md (hunk 7 above)
Comment (adk-swisstopo) [reply TERRAFORM.md:19]: AFAICT Terraform doesn't have a concept of "project" and a module is just any directory that has .tf files in it. Some modules are "root modules" (i.e. you can run "terraform" on that directory), all the others are "child modules" (or perhaps "orphaned modules" in some cases). Some child modules are reusable. The way I have heard you use these words is a bit different :) I think I can live with any reasonable definition/nomenclature but I would really like we nail one down in this document and my preference it to reuse as much of upstream as possible (and we may still need to define other words/concepts on top of that of course).

https://developer.hashicorp.com/terraform/language/modules

File: TERRAFORM.md (hunk 6 above)
Comment (ltshb) [reply TERRAFORM.md:39]: Good idea.

Actually we do have a way to check, we have our `tf-all` zsh function that gives you the possiblity to run `terraform plan` over the whole repository so you can check the whole diff in one command and/or apply it in one command. But this `tf-all` is a function that we implemented that you can find here https://github.com/geoadmin/infra-ansible-bgdi-collection

File: TERRAFORM.md (hunk 8 above)
Comment (ltshb) [reply TERRAFORM.md:52]: With some bash magic we could but this would be complicated in my opinion and I would not try to do it as it might add more trouble than it problem solve.

File: TERRAFORM.md (hunk 11 above)
Comment (ltshb) [reply TERRAFORM.md:90]: Here we could solve all this by using Atlantis https://www.runatlantis.io/ but it was decided that for the moment we don't want another tool to manage

File: TERRAFORM.md (hunk 5 above)
Comment (adk-swisstopo) [reply TERRAFORM.md:35]: Fair enough about enforcement but then I suggest explaining in more details what "keep it small" means, maybe with examples. Like you did in this review comment.

For the structure, what about a link to Confluence? I understand external readers won't be able to click it but the audience for this documentation explaining how our Terraform config works is internal anyway since external people don't have access to that either.

File: TERRAFORM.md (hunk 15 above)
Comment (ltshb) [reply TERRAFORM.md:103]: Unfortunately we don't have any CI yet for terraform, but this would be great to have one that at least check formatting as this is very easy to do with `terraform fmt` and `git status --pocelain` commands

File: TERRAFORM.md (hunk 13 above)
Comment (ltshb) [reply TERRAFORM.md:96]: Unfortunately not, or I'm not aware how.

File: TERRAFORM.md (hunk 12 above)
Comment (ltshb) [reply TERRAFORM.md:93]: Removing it is ok when it don't make sense

File: CI.md (hunk 4 above)
Comment (ltshb) [reply CI.md:25]: I added the link. 
Currently we don't have README linter, which would be good for README formatting but not for such simpler case in my opinion.

File: TERRAFORM.md (hunk 7 above)
Comment (ltshb) [reply TERRAFORM.md:19]: I propose to use
- Root Module := for terraform module that we can apply, what I used to call project
- Child Module := for non appliable module, that are tight to a root module (without backend, provider definition) (currently we don't have any of those.)
  - Reusable Module := for module non appliable, that can be used in any root module
@adk-swisstopo what do you think ?

File: TERRAFORM.md (hunk 9 above)
Comment (ltshb) [reply TERRAFORM.md:58]: Same I would not try to add a complex CI to enforce everything

File: TERRAFORM.md (hunk 15 above)
Comment (ltshb) [reply TERRAFORM.md:103]: Created ticket PB-608

**Review by ltclm — APPROVED** [review]

File: TERRAFORM.md (hunk 7 above)
Comment (adk-swisstopo) [reply TERRAFORM.md:21]: Sounds good and I think that matches the upstream definitions. I suggest adding something along the lines of "We use the terms 'module', 'root module' and 'reusable module' in the same way than [the Terraform documentation](https://developer.hashicorp.com/terraform/language/modules)." Maybe also "Historically, we have sometimes used the word 'project' to mean 'root module'."

File: TERRAFORM.md [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+```text
+|- modules/                   # terraform reusable modules for the whole repository
+|- ORGANIZATIONS_OR_ACCOUNT   # e.g. AWS Account name
+|           |- modules        # terraform reusable modules for this organization/account
+|           |- shared         # shared resources
+|           |- ENVIRONMENT    # resource pro environment
+```
+
+Then it should mirror the hierarchical structure of the infrastructure (see [PP BGDI Infra Architecture](https://ltwiki.adr.admin.ch:8443/display/PB/Runtime+Infrastructure+Stack)). Note that each terraform module (project) should be kept small.
```
Comment (adk-swisstopo) [inline TERRAFORM.md:37]: Remove the "(project)" :) Unless you mean "root module"? But in this case I think it's really just "module".

File: TERRAFORM.md (hunk 8 above)
Comment (adk-swisstopo) [reply TERRAFORM.md:46]: I think it might be useful to enforce this and other conventions in this way but the tools for parsing terraform config seem rather limited at the moment. The supported way is apparently terraform-config-inspect which doesn't even bother surfacing the backend definition. I agree hacking some bash script is likely going to be too fragile to be worth it.

File: TERRAFORM.md (hunk 11 above)
Comment (adk-swisstopo) [reply TERRAFORM.md:110]: Sure. But for now can we explain here why it's important to open a PR showing the intent of the change before applying the change? I think it's to allow other to easily track down what's happening in case of inconsistency/conflict but there may be other reasons?

File: TERRAFORM.md (hunk 15 above)
Comment (adk-swisstopo) [reply TERRAFORM.md:119]: Thank you. Would it make sense to mention that ticket here?

File: TERRAFORM.md (hunk 6 above)
Comment (adk-swisstopo) [reply TERRAFORM.md]: Awesome. Should we call this out in this documentation? Or even better if we can make the CI show the output of "tf-all plan" on every PR :)

File: TERRAFORM.md (hunk 7 above)
Comment (adk-swisstopo) [reply TERRAFORM.md:21]: You did that already and I am just confused about the github review workflow. Thank you.

File: TERRAFORM.md (hunk 6 above)
Comment (ltshb) [reply TERRAFORM.md]: For the CI I would not use tf-all, there is much better tool like atlantis for this, so if we want CI to do more than just checking formatting I would first consider such tool before trying to implement something ourself.
I already thought moving the `tf-all` command outside the infra-ansible-bgdi-collection to make it public and provide it as a oh-my-zsh plugin, this way the community could also use it and we could mention it here, but this would require some work.
